### PR TITLE
Implement set_maximized, get_current_monitor, set_fullscreen and set_decorations for windows

### PR DESCRIPTION
**Pull request by edwin0cheng**

### CODE REVIEW

**Comment by edwin0cheng** [comment]
Should fixed part of the https://github.com/tomaka/winit/issues/72

**Comment by francesca64** [comment]
Thanks so much for this! This will be useful for a lot of people.

I won't have time to give this a full review until later today, though what I've looked at so far looks good. However, this needs a CHANGELOG entry.

**Comment by edwin0cheng** [comment]
Just a minor bug first for rust-stable (I use nightly) and added support for with_maximized. (So that with_decorations, with_fullscreen all works now). 

It should be completed. Thanks :)

**Comment by francesca64** [comment]
After testing this on Windows 10, everything works as expected. However, I've succeeded in breaking things: while toggling either decorations or fullscreen *rapidly* repeatedly, the window in question eventually freezes. My other windows don't freeze, continuing to play their animations, but as soon as one of those windows is focused (in other words, as soon as it interacts with the backend), it freezes as well. This can happen using your updated fullscreen example program, though it's far easier to reproduce if you change the key handlers to respond to `Pressed` rather than `Released`, since then you can simply hold down the key and encounter the problem very readily. This fortunately isn't an issue with maximization.

**Comment by edwin0cheng** [comment]
@francesca64 I changed the fullscreen example use Pressed now. I just tested this and it should works now.

By the way, the cause of freezes are related to [this block](https://github.com/tomaka/winit/blob/master/src/platform/windows/events_loop.rs#L407), which i didn't fully understand why it would write in this way..

File: src/platform/windows/events_loop.rs [hunk 1]
```diff
@@ -273,6 +286,23 @@ impl EventsLoopProxy {
             }
         }
     }
+
+    pub fn execute_in_thread<F>(&self, function: F)
+        where F: FnMut(Inserter) + Send + 'static
+    {
+        unsafe {
+            let boxed = Box::new(function) as Box<FnMut(_)>;
+            let boxed2 = Box::new(boxed);
```
[the remark this answers is not in the document]
Comment (edwin0cheng) [reply src/platform/windows/events_loop.rs]: This execute_in_thread function i just copy from [EventsLoop::execute_in_thread](https://github.com/edwin0cheng/winit/blob/bd33cef0f65ee98b4e5785cc977d2972ba983edb/src/platform/windows/events_loop.rs#L233). Maybe i should copy the comment as well? Or i should refactor it out as a common implementation function ?

[the remark this answers is not in the document]
Comment (francesca64) [reply src/platform/windows/events_loop.rs]: Refactoring would be lovely.

**Comment by edwin0cheng** [comment]
Done ^^

**Comment by edwin0cheng** [comment]
Good catch, but im away from keyboard now, would check it tonight(asia time), thanks

**Comment by edwin0cheng** [comment]
@francesca64 It should be fixed :)

**Comment by francesca64** [comment]
Great, I've confirmed that it's fixed.

Here's another bug:
1. Run fullscreen example program
2. Exit fullscreen
3. Press key to maximize
4. Enter fullscreen
5. Press key to un-maximize
6. Exit fullscreen
7. (Window will still be maximized) press key to un-maximize; nothing happens until the second press

This fortunately doesn't happen for decoration toggling.

Using the built-in maximization button causes things like this to happen:
1. Run fullscreen example program
2. Exit fullscreen
3. Press the window's maximization button
4. Press key to un-maximize; nothing happens until the second press

It's also important to account for the user un-maximizing the window by moving it:
1. Run fullscreen example program
2. Exit fullscreen
3. Press key to maximize
4. Move window to un-maximize
5. Press key to maximize; nothing happens until the second press

**Comment by edwin0cheng** [comment]
> Using the built-in maximization button causes things like this to happen:
> 
>     Run fullscreen example program
>     Exit fullscreen
>     Press the window's maximization button
>     Press key to un-maximize; nothing happens until the second press
> 
> It's also important to account for the user un-maximizing the window by moving it:
> 
>     Run fullscreen example program
>     Exit fullscreen
>     Press key to maximize
>     Move window to un-maximize
>     Press key to maximize; nothing happens until the second press

It is because the fullscreen example just use a bool to indicate Maximize status, which is not sync to the system status.

**Comment by edwin0cheng** [comment]
> 
>     Run fullscreen example program
>     Exit fullscreen
>     Press key to maximize
>     Enter fullscreen
>     Press key to un-maximize
>     Exit fullscreen
>     (Window will still be maximized) press key to un-maximize; nothing happens until the second press
> 
https://github.com/tomaka/winit/pull/457/commits/9f52ed1d453cfd59867ba1cee42f78fe488691cb should fixed this

**Comment by francesca64** [comment]
Right, good point about it being an issue in the example and not your implementation, sorry about that.

While it does fix my first issue, there are still issues with the wrong state being restored after using the built-in maximization button or moving to un-maximize.
1. Run fullscreen example program
2. Exit fullscreen
3. Press the window's maximization button
4. Enter fullscreen
5. Exit fullscreen
6. The window is no longer maximized

and:
1. Run fullscreen example program
2. Press key to maximize
3. Exit fullscreen
3. Move window to un-maximize
4. Enter fullscreen
5. Exit fullscreen
6. The window is maximized

**Comment by edwin0cheng** [comment]
Sorry about that, the maximized state is a little bit tricky to implement correctly.  Anyway, it should be fixed all the issues listed.

**Comment by francesca64** [comment]
Awesome! This is the last thing, I promise: in the CHANGELOG, you accidentally put `WindowBuilder::set_maximized` instead of `WindowBuilder::with_maximized`.

**Comment by edwin0cheng** [comment]
Sorry about that, thank you so much to help me to catch out these bugs. It will be my second closed PR :)

**Comment by francesca64** [comment]
No problem! This PR is very much appreciated.

**Pull request merged**